### Imports

In [35]:
import pandas as pd 
import numpy as np 
import textwrap
from sklearn.preprocessing import LabelEncoder




#### Data loading


In [7]:
data_raw = pd.read_csv('../data/Evaluation.csv', sep=',')

In [27]:
print(data_raw.shape)


(6915, 10)


In [9]:
data_raw.head()

,Unnamed: 0,CaseSummary,Disease,Synonyms,ICD-10,ICD-11,OMIM,GARD,UMLS
0,26357,"At the time of the follow-ups, 1, 2, and 6 mon...",Waldenström macroglobulinemia,NaN,C88.0,2A85.4,153600; 610430,7872.0,C0024419
1,42784,"After transferring to our Center, the parents ...",Cystic fibrosis,CF; Mucoviscidosis,E84.0; E84.1; E84.8,CA25,219700,6233.0,C0010674
2,5807,A woman in her forties visited a local obstetr...,Neurofibromatosis type 1,Nonmosaic neurofibromatosis type 1; Von Reckli...,Q85.0,LD2D.10,162200; 162210; 613675,7866.0,C0027831
3,16473,"Although many cases have been published, it is...",Interatrial communication,ASD; Atrial septal defect; Interauricular comm...,Q21.1,LA8E.Y,108800; 607941; 611363; 612794; 613087; 614089...,NaN,C0018817
4,62585,"A 63-year-old man, residing in the Var county ...",Rickettsialpox,NaN,A79.1,1C32,NaN,NaN,C0035597


In [11]:
data_raw.columns

Index(['Unnamed: 0', 'CaseSummary', 'Disease', 'Synonyms', 'ICD-10', 'ICD-11',
       'OMIM', 'GARD', 'UMLS'],
      dtype='object')

In [30]:
diagnosis_len = data_raw['Disease'].nunique()
print(f'Number of unique diagnoses: {diagnosis_len}')

Number of unique diagnoses: 1012


In [19]:
data_raw[['Disease', 'CaseSummary']][0:1]

,Disease,CaseSummary
0,Waldenström macroglobulinemia,"At the time of the follow-ups, 1, 2, and 6 mon..."


In [20]:
data_raw["text_length"] = data_raw["CaseSummary"].apply(len)

In [25]:
text = data_raw['CaseSummary'].iloc[0]
print(textwrap.fill(text, width=80))

At the time of the follow-ups, 1, 2, and 6 months after the injection, BCVA
remained stable at 20/30, intraocular pressure was 14 mm Hg, and the patient did
not show any sign of opacity of the lens or ocular inflammation.


In [26]:
data_raw.describe()

,Unnamed: 0,GARD,text_length
count,6915.000000,5412.000000,6915.000000
mean,34620.735792,6396.631744,2328.567028
std,19942.879426,2996.250120,1959.878098
min,21.000000,3.000000,6.000000
25%,17287.000000,5120.000000,863.000000
50%,34850.000000,6851.000000,1776.000000
75%,51712.500000,7887.000000,3345.000000
max,69141.000000,13643.000000,44156.000000


In [28]:
print(data_raw.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6915 entries, 0 to 6914
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   6915 non-null   int64  
 1   CaseSummary  6915 non-null   object 
 2   Disease      6915 non-null   object 
 3   Synonyms     4845 non-null   object 
 4   ICD-10       6235 non-null   object 
 5   ICD-11       6218 non-null   object 
 6   OMIM         4537 non-null   object 
 7   GARD         5412 non-null   float64
 8   UMLS         6897 non-null   object 
 9   text_length  6915 non-null   int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 540.4+ KB
None


In [29]:
data_raw.isnull().sum()

Unnamed: 0        0
CaseSummary       0
Disease           0
Synonyms       2070
ICD-10          680
ICD-11          697
OMIM           2378
GARD           1503
UMLS             18
text_length       0
dtype: int64

In [31]:
print(data_raw[data_raw['CaseSummary'].str.len() == 6]['CaseSummary'])

2371    Case 1
5524    Case 1
Name: CaseSummary, dtype: object


In [32]:
data_raw.columns

Index(['Unnamed: 0', 'CaseSummary', 'Disease', 'Synonyms', 'ICD-10', 'ICD-11',
       'OMIM', 'GARD', 'UMLS', 'text_length'],
      dtype='object')

In [33]:
data = data_raw[['CaseSummary', 'Disease', 'text_length', 'Synonyms']].copy()

In [34]:
data.head()

,CaseSummary,Disease,text_length,Synonyms
0,"At the time of the follow-ups, 1, 2, and 6 mon...",Waldenström macroglobulinemia,221,NaN
1,"After transferring to our Center, the parents ...",Cystic fibrosis,1239,CF; Mucoviscidosis
2,A woman in her forties visited a local obstetr...,Neurofibromatosis type 1,309,Nonmosaic neurofibromatosis type 1; Von Reckli...
3,"Although many cases have been published, it is...",Interatrial communication,658,ASD; Atrial septal defect; Interauricular comm...
4,"A 63-year-old man, residing in the Var county ...",Rickettsialpox,1642,NaN


In [36]:
le = LabelEncoder()
data['category_encoded'] = le.fit_transform(data['Disease'])

# Show the mapping
label_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print("Label mapping:")
# for cat, code in sorted(label_mapping.items(), key=lambda x: x[1]):
#     print(f"  {code} -> {cat}")

# print()
# data[['Disease', 'category_encoded']].head(10)

Label mapping:


In [37]:
data.head()

,CaseSummary,Disease,text_length,Synonyms,category_encoded
0,"At the time of the follow-ups, 1, 2, and 6 mon...",Waldenström macroglobulinemia,221,NaN,975
1,"After transferring to our Center, the parents ...",Cystic fibrosis,1239,CF; Mucoviscidosis,249
2,A woman in her forties visited a local obstetr...,Neurofibromatosis type 1,309,Nonmosaic neurofibromatosis type 1; Von Reckli...,681
3,"Although many cases have been published, it is...",Interatrial communication,658,ASD; Atrial septal defect; Interauricular comm...,492
4,"A 63-year-old man, residing in the Var county ...",Rickettsialpox,1642,NaN,838


In [38]:
data['category_encoded'].nunique()

1012

In [39]:
data.isnull().sum()

CaseSummary            0
Disease                0
text_length            0
Synonyms            2070
category_encoded       0
dtype: int64

In [40]:
data[['Synonyms']].iloc[0:5]

,Synonyms
0,NaN
1,CF; Mucoviscidosis
2,Nonmosaic neurofibromatosis type 1; Von Reckli...
3,ASD; Atrial septal defect; Interauricular comm...
4,NaN


In [41]:
data['Synonyms'] = data['Synonyms'].fillna("Not available")

In [42]:
data[['Synonyms']].iloc[0:5]

,Synonyms
0,Not available
1,CF; Mucoviscidosis
2,Nonmosaic neurofibromatosis type 1; Von Reckli...
3,ASD; Atrial septal defect; Interauricular comm...
4,Not available


In [43]:
data.isnull().sum()

CaseSummary         0
Disease             0
text_length         0
Synonyms            0
category_encoded    0
dtype: int64

In [45]:
data.to_csv('../data/cleaned_data.csv', index=False)

### Summary

**Dataset:** `Evaluation.csv` — 6,915 clinical case summaries spanning **1,012 unique rare diseases**.

**Exploration performed:**
- Loaded and inspected shape, columns, dtypes, and basic statistics
- Checked missing values across all columns — `Synonyms` (2,070), `OMIM` (2,378), `GARD` (1,503) had the most nulls; `CaseSummary` and `Disease` had zero
- Computed `text_length` for each case summary (range: 6–44,156 chars, median ~1,776)
- Identified the shortest summaries (6 chars) as placeholder text ("Case 1")
- Examined disease synonym coverage

**Key findings:**
1. The dataset is highly multi-class: 1,012 disease labels across 6,915 samples (~6.8 samples per disease on average)
2. Case summaries vary widely in length (6 to 44K characters), indicating heterogeneous data quality
3. ~30% of rows lack synonyms; medical coding columns (ICD-10, ICD-11, OMIM, GARD) also have notable missingness
4. `CaseSummary` and `Disease` are fully populated — no nulls

**Cleaning steps:**
- Selected a subset of columns for downstream use
- Filled missing `Synonyms` with `"Not available"`
- Label-encoded the `Disease` column into `category_encoded`

**Columns in `cleaned_data.csv`:**

| Column | Description |
|---|---|
| `CaseSummary` | Clinical case narrative (text) |
| `Disease` | Rare disease name (target label) |
| `text_length` | Character length of `CaseSummary` |
| `Synonyms` | Disease synonyms (nulls filled with "Not available") |
| `category_encoded` | Integer-encoded disease label (0–1011) |